xformers/components/attention/csrc/cpu/attention.cpp

```diff
@@ -0,0 +1,151 @@
+#include <ATen/ATen.h>
+#include <ATen/Parallel.h>
+#include <torch/library.h>
+#include <cmath>
+#include <vector>
+
+#include <ATen/cpu/vec/functional.h>
+#include <ATen/cpu/vec/vec.h>
+
+namespace {
+
+template <typename scalar_t>
+void fill_zero(scalar_t* buf, int64_t K) {
+  for (int64_t k = 0; k < K; k++) {
+    buf[k] = 0;
+  }
+}
+
+template <typename scalar_t, int K>
+scalar_t max(scalar_t* buf) {
+  scalar_t m = buf[0];
+  for (int64_t k = 1; k < K; k++) {
+    m = buf[k] > m ? buf[k] : m;
+  }
+  return m;
+}
+
+template <typename scalar_t>
+void attention_kernel(
+    at::TensorAccessor<scalar_t, 3> output,
+    at::TensorAccessor<scalar_t, 3> query,
+    at::TensorAccessor<scalar_t, 3> key,
+    at::TensorAccessor<scalar_t, 3> value,
+    at::TensorAccessor<scalar_t, 3> buffer //,
+    // at::TensorAccessor<int64_t, 2> mask
+) {
+  constexpr int64_t BLOCK = 1; // 8;
```

Comment:
I'm guessing that there's some speed to gain here, to remove some reads / reuse them across a couple of rows ?

Reply:
actually this is the size of the fetch over N, my bad, I'm guessing (hoping) that the compiler groups them automatically

Reply:
Exactly. Moving from 1 -> 8 brought a significant speedup. I moved it back to 1 before sending the PR because I was a bit lazy and didn't want to bother handling the remainder cases like I do in the GPU code.

That being said, now that I performed unrolling of both dimensions in the CUDA kernel, I could probably copy-paste the CUDA code and change a few things in the hope of making the CPU code faster.

But given that CPU is normally used for prototyping most of the time, I didn't prioritize it further.

Reply:
makes sense, as it is the memory accesses are a little too granular, but not a big issue. I think it's fine not to get it super duper optimized right now, but we could add a comment about it for the future, out of context, us ?